src/internal/modules/agreements/controller.js

```diff
@@ -30,5 +35,73 @@ const postDeleteAgreement = async (request, h) => {
   return h.redirect(`/licences/${document.document_id}#charge`);
 };
 
+const getEndAgreement = async (request, h) => {
+  const { agreement, licence, document } = request.pre;
+  const { agreementId } = request.params;
+  const { endDate } = helpers.sessionManager(request, agreementId);
+  return h.view('nunjucks/form', {
+    ...getDefaultView(request),
+    pageTitle: 'Set agreement end date',
+    back: `/licences/${document.document_id}#charge`,
+    agreement,
+    licenceId: licence.id,
+    form: sessionForms.get(request, endAgreementForm(request, endDate))
+  });
+};
+
+const postEndAgreement = async (request, h) => {
+  const { licenceId, agreementId } = request.params;
+  const formEndDate = new Date(`${request.payload['endDate-year']}-${request.payload['endDate-month']}-${request.payload['endDate-day']}`);
+
+  const form = await forms.handleRequest(endAgreementForm(request, formEndDate), request, endAgreementFormSchema(request, h));
+
+  const goBack = () => {
+    return h.postRedirectGet(form, `/licences/${licenceId}/agreements/${agreementId}/end`);
+  };
+  console.log(form);
+  if (form.isValid) {
+    try {
+      await helpers.sessionManager(request, agreementId, { endDate: formEndDate });
+      return h.redirect(`/licences/${licenceId}/agreements/${agreementId}/end/confirm`);
+    } catch (err) {
+      return goBack();
+    }
+  } else {
+    return goBack();
+  }
+};
+
+const getConfirmEndAgreement = async (request, h) => {
+  const { agreement, licence, document } = request.pre;
+  const { agreementId } = request.params;
+  const { endDate } = await helpers.sessionManager(request, agreementId);
+  return h.view('nunjucks/agreements/end', {
+    ...getDefaultView(request),
+    pageTitle: 'You\'re about to end this agreement',
+    back: `/licences/${document.document_id}#charge`,
```

Comment:
I think this should be going back to the previous page, where the user enters the end date, as opposed to the charge info tab

Reply:
Fair point. I'll change it